testing/test_assertion.py

```diff
@@ -572,6 +573,111 @@ def test_fmt_multi_newline_before_where(self):
+        assert result != expl
+        assert len(result) == 8 + self.LINES_IN_TRUNCATION_MSG
+        assert "Full output truncated" in result[-1]
+        assert "93 lines hidden" in result[-1]
+        last_line_before_trunc_msg = result[- self.LINES_IN_TRUNCATION_MSG -1]
+        assert last_line_before_trunc_msg.endswith("...")
+
+    def test_truncates_at_8_lines_when_first_8_lines_are_EQ_max_chars(self):
+        expl = ['a' * 80 for x in range(16)]
+        result = truncate._truncate_explanation(expl, max_lines=8, max_chars=8*80)
+        assert result != expl
+        assert len(result) == 8 + self.LINES_IN_TRUNCATION_MSG
+        assert "Full output truncated" in result[-1]
+        assert "9 lines hidden" in result[-1]
+        last_line_before_trunc_msg = result[- self.LINES_IN_TRUNCATION_MSG -1]
+        assert last_line_before_trunc_msg.endswith("...")
+
+    def test_truncates_at_4_lines_when_first_4_lines_are_GT_max_chars(self):
+        expl = ['a' * 250 for x in range(10)]
+        result = truncate._truncate_explanation(expl, max_lines=8, max_chars=999)
+        assert result != expl
+        assert len(result) == 4 + self.LINES_IN_TRUNCATION_MSG
+        assert "Full output truncated" in result[-1]
+        assert "7 lines hidden" in result[-1]
+        last_line_before_trunc_msg = result[- self.LINES_IN_TRUNCATION_MSG -1]
+        assert last_line_before_trunc_msg.endswith("...")
+
+    def test_truncates_at_1_line_when_first_line_is_GT_max_chars(self):
+        expl = ['a' * 250 for x in range(1000)]
+        result = truncate._truncate_explanation(expl, max_lines=8, max_chars=100)
+        assert result != expl
+        assert len(result) == 1 + self.LINES_IN_TRUNCATION_MSG
+        assert "Full output truncated" in result[-1]
+        assert "1000 lines hidden" in result[-1]
+        last_line_before_trunc_msg = result[- self.LINES_IN_TRUNCATION_MSG -1]
+        assert last_line_before_trunc_msg.endswith("...")
+
+    def test_full_output_truncated(self, monkeypatch, testdir):
+        """ Test against full runpytest() output. """
+
+        line_count = 7
+        line_len = 100
+        expected_truncated_lines = 2
+        testdir.makepyfile(r"""
+            def test_many_lines():
+                a = list([str(i)[0] * %d for i in range(%d)])
+                b = a[::2]
+                a = '\n'.join(map(str, a))
+                b = '\n'.join(map(str, b))
+                assert a == b
+        """ % (line_len, line_count))
+        monkeypatch.delenv('CI', raising=False)
+
+        result = testdir.runpytest()
+        # without -vv, truncate the message showing a few diff lines only
+        result.stdout.fnmatch_lines([
+            "*- 1*",
+            "*- 3*",
+            "*- 5*",
+            "*truncated (%d lines hidden)*use*-vv*" % expected_truncated_lines,
+        ])
+
+        result = testdir.runpytest('-vv')
+        result.stdout.fnmatch_lines([
```

Comment:
The output from this `runpytest` call is:

```
    def test_many_lines():
        a = list([str(i)[0] * 100 for i in range(7)])
        b = a[::2]
        a = '\n'.join(map(str, a))
        b = '\n'.join(map(str, b))
>       assert a == b
E       assert '000000000000...6666666666666' == '0000000000000...6666666666666'
E           0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
E         - 1111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
E           2222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222
E         - 3333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333
E           4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
E         - 5555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555
E           6666666666666666666666666666666666666666666666666666666666666666666666666666666666666666666666666666

test_full_output_truncated.py:6: AssertionError
========================== 1 failed in 0.01 seconds ===========================
```

Because `"*- 5*` will match even when truncating the output as the match above this one show, I think this `fnmatch_lines` call and the next should match against the `6*` line to demonstrate the the output is not being truncated, something like:

```python
result.stdout.fnmatch_lines([
    "*  6*",
])
```